spyder/app/tests/test_mainwindow.py

```diff
@@ -530,6 +535,74 @@ def test_single_instance_and_edit_magic(main_window, qtbot, tmpdir):
     main_window.editor.close_file()
 
 
+@pytest.mark.slow
+@pytest.mark.use_introspection
+def test_leaks(main_window, qtbot):
+    """
+    Test leaks in mainwindow when closing a file or a console.
+
+    Many other ways of leaking exist but are not covered here.
+    """
+    # Wait until the window is fully up
+    shell = main_window.ipyconsole.get_current_shellwidget()
+    qtbot.waitUntil(
+        lambda: shell._prompt_html is not None, timeout=SHELL_TIMEOUT)
+
+    # Count initial objects
+    # Only one of each should be present, but because of many leaks,
+    # this is most likely not the case. Here only closing is tested
+    shell.wait_all_shutdown()
+    gc.collect()
+    objects = gc.get_objects()
+    n_code_editor_init = 0
+    for o in objects:
+        if type(o).__name__ == "CodeEditor":
+            n_code_editor_init += 1
+    n_shell_init = 0
+    for o in objects:
+        if type(o).__name__ == "ShellWidget":
+            n_shell_init += 1
+    del objects
+
+    # Open a second file and console
+    main_window.editor.new()
+    main_window.ipyconsole.create_new_client()
+    # Do something interesting in the new window
+    code_editor = main_window.editor.get_focus_widget()
+    # Show an error in the editor
+    code_editor.set_text("aaa")
+    del code_editor
+
+    shell = main_window.ipyconsole.get_current_shellwidget()
+    qtbot.waitUntil(
+        lambda: shell._prompt_html is not None, timeout=SHELL_TIMEOUT)
```

Comment:
I think this is not necessary because you already waited for the console to be up at the beginning of the test, right?

Reply:
Well here I am waiting for it to restart? Is it not how you should do it?

Reply:
I just created a new client, so I am waiting for it to come online

Reply:
Ok, sorry, I missed that you did that some lines above.